### ISSUE TITLE
[4][com_finder][ACL bypass] Smart Search reveals author names of unpublished/acl restricted articles.

### ISSUE THREAD

**Issue by PhilETaylor** [joomla/joomla-cms#32902]
### Steps to reproduce the issue
tested on Joomla 4.0-dev

Create a menu link to Smart Search 
Visit that menu link and look in the Advanced Search -> Search by author dropdown - note what you see

Create a new UNPUBLISHED Article. Enter an Author Alias you would recognise. NOTE THE ARTICLE IS UNPUBLISHED. 

Visit that menu link and look in the Advanced Search -> Search by author dropdown - note what you see

### Expected result

I expect to NOT see the Author Alias of an unpublished item, if there are no published items with that same Author name. 

### Actual result

Smart search is leaking information and displaying the names of Authors of unpublished items, where there are zero published items by that author. 

### Also

repeat this with setting the article to an ACL level your public has no access to (like Special or Super Users). Repeat the test. You can now view the author name of items that are restricted to you by ACL. 

**This is probably a security issue then as its ACL not being applied correctly too.** 

@joomla/security

**Comment by brianteeman** [comment]
We addressed this years ago in j3. Finder in J4 is an almost complete rewrite allegedly

**Comment by PhilETaylor** [comment]
Im no com_finder expert.. but storing the full route to items in the db and a full serialised php object per item seems overkill. my finder_links table is over 200Mb in size!

**Comment by Hackwar** [comment]
@brianteeman No, we did not address this years ago. We adressed, that words and small phrases could be leaked via the search suggestions. Smart Search in J4 also isn't a rewrite, but incremental enhancements of what we got in J2.5. While none of the features are rocket science, I wouldn't have been able to come up with the combination of all of this and the code is in some cases very different to what I would write. That at the same time also makes it clear to me, that this has also not been developed from the ground up by those who donated it to Joomla. From the coding style, the overall construction, it is pretty clear that this originated as a standalone project somewhere outside of Joomla by one group and was then adapted to our CMS by another group of people. Long story short: I improved it, but didn't rewrite it. Most changes were natural improvements. Some of the biggest changes unfortunately weren't backwards compatible, which is why this is done in J4 and not already in J3.

@PhilETaylor The issue is also present in J3. I don't see a way to prevent this, except for adding parameters in the finder plugins to not store these taxonomies.
The data stored in the finder_links table is indeed a full copy of the sites content, similar to the UCM table. I don't like it either, but I also wouldn't know a different solution to this. At least none that is somewhere as performant as this. If you have a proposal how to better implement this, I'm all ears. We would need it for the tags rewrite at some point as well.

And while writing all of this, I actually do have an idea what to do. I'll post a PR soon.
